backend/sqlite/sqlite.go

```diff
@@ -455,26 +505,74 @@ func (be *sqliteBackend) createOrchestrationInstanceInternal(ctx context.Context
 		) VALUES (?, ?, ?, ?, ?, ?, ?)`,
 		startEvent.Name,
 		startEvent.Version.GetValue(),
-		startEvent.OrchestrationInstance.InstanceId,
+		instanceID,
 		startEvent.OrchestrationInstance.ExecutionId.GetValue(),
 		startEvent.Input.GetValue(),
 		"PENDING",
 		e.Timestamp.AsTime(),
 	)
 	if err != nil {
-		return fmt.Errorf("failed to insert into [Instances] table: %w", err)
+		return "", "", fmt.Errorf("failed to insert into [Instances] table: %w", err)
 	}
 
 	rows, err := res.RowsAffected()
 	if err != nil {
-		return fmt.Errorf("failed to count the rows affected: %w", err)
+		return "", "", fmt.Errorf("failed to count the rows affected: %w", err)
 	}
 
 	if rows <= 0 {
-		return backend.ErrDuplicateEvent
+		// query RuntimeStatus for the existing instance
+		queryRow := tx.QueryRowContext(
+			ctx,
+			`SELECT [RuntimeStatus] FROM Instances WHERE [InstanceID] = ?`,
+			startEvent.OrchestrationInstance.InstanceId,
+		)
+		var runtimeStatus *string
+		err = queryRow.Scan(&runtimeStatus)
+		if errors.Is(err, sql.ErrNoRows) {
+			return "", "", api.ErrInstanceNotFound
+		} else if err != nil {
+			return "", "", fmt.Errorf("failed to scan the Instances table result: %w", err)
+		}
+		return *runtimeStatus, instanceID, backend.ErrDuplicateEvent
+	}
+	return "", instanceID, nil
+}
+
+func (be *sqliteBackend) cleanupOrchestrationStateInternal(ctx context.Context, tx *sql.Tx, id api.InstanceID) error {
+	row := tx.QueryRowContext(ctx, "SELECT 1 FROM Instances WHERE [InstanceID] = ?", string(id))
+	if err := row.Err(); err != nil {
+		return fmt.Errorf("failed to query for instance existence: %w", err)
 	}
 
-	*instanceID = startEvent.OrchestrationInstance.InstanceId
+	var unused int
+	if err := row.Scan(&unused); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return api.ErrInstanceNotFound
+		} else {
+			return fmt.Errorf("failed to scan instance existence: %w", err)
+		}
+	}
+
+	_, err := tx.ExecContext(ctx, "DELETE FROM Instances WHERE [InstanceID] = ?", string(id))
+	if err != nil {
+		return fmt.Errorf("failed to delete from the Instances table: %w", err)
+	}
+
+	_, err = tx.ExecContext(ctx, "DELETE FROM History WHERE [InstanceID] = ?", string(id))
+	if err != nil {
+		return fmt.Errorf("failed to delete from History table: %w", err)
+	}
+
+	_, err = tx.ExecContext(ctx, "DELETE FROM NewEvents WHERE [InstanceID] = ?", string(id))
+	if err != nil {
+		return fmt.Errorf("failed to delete from NewEvents table: %w", err)
+	}
+
+	_, err = tx.ExecContext(ctx, "DELETE FROM NewTasks WHERE [InstanceID] = ?", string(id))
```

Comment:
It seems like there might be a lot of duplication between this and purge. Could those two operations share some of the implementation?